keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1197 @@
+kubelet plugin must double-check that the allocated resources are still
+available when NodePrepareResource is called. If not, the pod cannot
+start until the resource comes back. Treating this as a fatal error during
+pod admission would allow to delete the pod and trying again with a new
+one.
+
+### Using semantic parameters as claim parameters
+
+The author of the `cards.dra.example.com` DRA driver in the example above might
+define a CRD like this:
+
+```
+kind: CardParameters
+apiVersion: dra.example.com/v1alpha1
+metadata:
+  name: my-parameters
+  namespace: user-namespace
+  uid: foobar-uid
+...
+spec:
+  minimumAPI: "v1.0.0"
+  supportsPowerSaving: true
+  requirements:
+    RAM: 10Gi
+    ComputeUnits: 2
+```
+
+Note that all fields can be fully validated, including value ranges that are
+specific to the underlying hardware.
+
+When a user creates an object of that type, the DRA driver control plane must
+convert it into an in-tree ResourceClaimParameters object because the scheduler
+doesn't understand `CardParameters`. That object then looks like this:
+
+```
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+
+metadata:
+  # This cannot be the same as my-parameters because parameter objects with a different
+  # type might also use it. Instead, the original object gets linked to below.
+  name: someArbitraryName
+  namespace: user-namespace
+
+generatedFrom:
+  name: my-parameters
+  kind: CardParameters
+  apiGroup: dra.example.com
+  uid: foobar-uid
+  generation: 10 # Ensures that only up-to-date generated parameters are used.
+
+parameters:
+- name: meta
+  metadata:
+    # A CELL expression with access to the labels, versions, and values defined for
+    # the node resource.
+    selector: versions["API"] >= "v1.0.0" && labels.has("SupportsPowerSaving")
+- name: RAM
+  counter:
+    required: 10000000
+- name: ComputeUnits
+  counter:
+    required: 2
+```
+
+The semantic is that all parameters must be satisfied by a single resource of
+the driver. Note that for the "pool of cards" example that makes it impossible
+to ask for two cards with no additional selection criteria and then get one
+from `PoolOfLargeCards` and one from a hypothetical second pool
+`PoolOfSmallCards` (not included in the example above).
+
+The semantic models define how to match the parameters against the node
+resources. For the `counter` model, the required amount must be available. Once
+allocated, the remaining amount can be used for other claims. The `metadata`
+model only returns a yes/no decision.
+
+Instead of defining a vendor CRD, DRA driver authors or administrators may also
+decide to allow users to create and reference ResourceClaimParameters directly.
+
+### Communicating allocation to the DRA driver
+
+The scheduler decides which resources to use for a claim and how much of
+them. It also needs to take a snapshot of the full class and claim parameter
```

Comment:
We should avoid having the scheduler watch arbitrary CRs.
A better approach would be for the third-party controller to write the parameters into the claim spec, and make it immutable after the first write (or even just immutable when the claim is bound).

Reply:
I'm not sure we can write it to the claim spec, but I think we can store it in the ClaimParameters object. See https://github.com/kubernetes/enhancements/pull/4384#discussion_r1480356891

Reply:
RawExtension for vendor configuration parameter in ResourceClaimParameters was added. Same for resource class parameters.